### PR TITLE
feat(engines): add toutiao search engine

### PR DESCRIPTION
**Pull request by hubutui**
Supports these search types: synthesis, information, atlas, weitoutiao, and video. Multiple engine instances can be configured via `toutiao_pd` and `toutiao_filter_vendor` settings. Time range filtering is supported for synthesis search.

### What does this PR do?

Adds a new search engine for [Toutiao](https://www.toutiao.com) (今日头条), ByteDance's content aggregation and search platform used in China.

**Features:**

- **Six engine instances** out of the box:
  | Instance | Search Type | Shortcut |
  |---|---|---|
  | toutiao | Synthesis (综合, web-wide) | `!toutiao` |
  | toutiao site | Synthesis (综合, Toutiao-only) | `!ttsite` |
  | toutiao news | Information (资讯) | `!ttnews` |
  | toutiao images | Atlas / Images (图片) | `!ttimg` |
  | toutiao weitoutiao | Weitoutiao (微头条, microblog) | `!ttwtt` |
  | toutiao videos | Video (视频) | `!ttvid` |

- **Time range filtering** on synthesis search (day / week / month / year).
- **ttwid cookie** pre-fetched from ByteDance's registration API, cached via `EngineCache` with a 1-hour TTL to reduce anti-bot detection.
- **Captcha detection** — raises `SearxEngineCaptchaException` when a slide verification challenge is encountered.
- **URL sanitization** — filters out `sslocal://` app deep-links and `preview_article` editor URLs, replacing them with proper public article links.
- **Result parsing** from JSON embedded in `<script type="application/json">` tags (HTML DOM parsing for atlas/images).

Xiaoshipin (小视频, short videos) is intentionally excluded — it requires full browser cookies and does not work with server-side requests.

### How to test this PR locally?

Add the engine instances to `settings.yml`:

```yaml
  - name: toutiao
    engine: toutiao
    shortcut: toutiao

  - name: toutiao site
    engine: toutiao
    shortcut: ttsite
    toutiao_filter_vendor: site

  - name: toutiao news
    engine: toutiao
    shortcut: ttnews
    categories: news
    toutiao_pd: information
    time_range_support: false

  - name: toutiao images
    engine: toutiao
    shortcut: ttimg
    categories: images
    toutiao_pd: atlas
    time_range_support: false

  - name: toutiao weitoutiao
    engine: toutiao
    shortcut: ttwtt
    categories: social media
    toutiao_pd: weitoutiao
    time_range_support: false

  - name: toutiao videos
    engine: toutiao
    shortcut: ttvid
    categories: videos
    toutiao_pd: video
    time_range_support: false
```

Then test with bang shortcuts:

```
!toutiao 小米
!ttsite 小米
!ttnews 小米
!ttimg 比亚迪
!ttwtt 比亚迪
!ttvid 小米
```

**Edge cases:** Some search tabs (video, weitoutiao) may intermittently trigger anti-bot verification depending on request frequency and source IP. Synthesis search is generally the most stable.

### Code of Conduct

[AI Policy]: https://github.com/searxng/searxng/blob/master/AI_POLICY.rst

- [x] **I hereby confirm that this PR conforms with the [AI Policy].**

  AI tools used: Claude Code (codex) — used for code generation, debugging, and testing of the toutiao engine module. All generated code has been reviewed, tested manually against live responses, and adjusted to conform with SearXNG engine conventions.

### CODE REVIEW

**Comment by return42** [comment]
Thanks a lot for your PR :+1: 

Please familiarize yourself with the rules and adjust the PR and the commits accordingly:

- [Development Quickstart Guide](https://docs.searxng.org/dev/quickstart.html)
- [How-to contribute](https://github.com/searxng/searxng/blob/master/CONTRIBUTING.rst#how-to-contribute)
- [Git Commits & Change Management](https://docs.searxng.org/dev/commits.html)
- [AI Policy](https://github.com/searxng/searxng/blob/master/AI_POLICY.rst)

By following the rules, you make the review process easier for us, allowing your PR to be merged more quickly.

**Comment by Bnyro** [comment]
> adjusted to conform with SearXNG engine conventions.
>
Okay, and what are these conventions, Claude?

As per our AI policy, you must write the full pull request description yourself without the use of AI, so please rewrite the PR description in your own words.

---

I had a quick look at the code, it still looks very unfinished and not polished - please go through the code and simplify it where possible first before we continue with reviewing.

File: searx/engines/toutiao.py [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+        query_params["cur_tab_title"] = "gallery"
+        query_params["action_type"] = "search_subtab_switch"
+    elif toutiao_pd == "weitoutiao":
+        query_params["from"] = "weitoutiao"
+        query_params["cur_tab_title"] = "weitoutiao"
+        query_params["action_type"] = "search_subtab_switch"
+    elif toutiao_pd == "video":
+        query_params["from"] = "video"
+        query_params["cur_tab_title"] = "video"
+        query_params["action_type"] = "search_subtab_switch"
```
Comment (Bnyro) [inline searx/engines/toutiao.py:161]: Please refactor this, you can just set them to `toutiao_pd` without doing any if/else

File: searx/engines/toutiao.py [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+    else:
+        # synthesis (综合): default search tab
+        query_params["from"] = "search_tab"
+        query_params["cur_tab_title"] = "search_tab"
+        query_params["action_type"] = "search_subtab_switch"
+        query_params["filter_vendor"] = toutiao_filter_vendor
+        query_params["index_resource"] = toutiao_filter_vendor
+
+    time_range = params.get("time_range")
```
Comment (Bnyro) [inline searx/engines/toutiao.py:170]:
```suggestion
    time_range = params["time_range"]
```
I think I already mentioned this in another of your PRs, time_range is always set (i.e. it is None for searching all time ranges)

File: searx/engines/toutiao.py [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+    url = (
+        display.get("info", {}).get("url", "")
+        or data.get("article_url", "")
+        or data.get("url", "")
+    )
+
+    # Reject sslocal:// deep-links and preview_article URLs so the
+    # source_url / group_id fallbacks below get a chance to work.
+    if _is_bad_url(url):
+        url = ""
```
Comment (Bnyro) [inline searx/engines/toutiao.py:532]: Why override method from above here?

**Comment by hubutui** [comment]
ok, I'll check it later. I tested these search engines again and found I got CAPTCHA very offen. Maybe they're not ready to added to SearXNG.

**Comment by return42** [comment]
I close this PR for now .. seems you ignored first comment from above https://github.com/searxng/searxng/pull/6342#issuecomment-4842981439

We can reopen the PR once all the rules are understood and the PR has been cleared of all the AI noise .. BTW: the code was generated, we basically don't review generated code.
